Add parsing for function pointer types

333fred opened the pull request:
@dotnet/roslyn-compiler @AlekseyTs for review.

Latest grammar: https://github.com/dotnet/csharplang/blob/master/proposals/function-pointers.md#function-pointer-syntax

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:
```diff
@@ -335,6 +336,14 @@ array_rank_specifier
   ;
 
+function_pointer_type
+  : 'delegate' '*' ('cdecl' | 'managed' | 'stdcall' | 'thiscall' | 'unmanaged')? '<' function_pointer_parameter_or_return_type (',' function_pointer_parameter_or_return_type)* '>'
```
makes me so happy to be able to see this :D

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:
```diff
@@ -335,6 +336,14 @@ array_rank_specifier
+  : 'delegate' '*' ('cdecl' | 'managed' | 'stdcall' | 'thiscall' | 'unmanaged')? '<' function_pointer_parameter_or_return_type (',' function_pointer_parameter_or_return_type)* '>'
+  ;
+
+function_pointer_parameter_or_return_type
```
maybe just `ModifiedType`.

333fred replied:
I wanted to be very specific with the grammar here because this can accept both return-type modifiers, like `ref readonly`, and parameter-type modifiers, like `in`.

---
In reply to: [348012458](https://github.com/dotnet/roslyn/pull/39879#discussion_r348012458) [](ancestors = 348012458)

CyrusNajmabadi replied:
I don't see the problem.  `ParameterSyntax` (and `ModifiedSyntax` if you went that route) would allow this.  My point is simply not to call it FunctionPointerParameterOrReturnType.  It's really unnecessarily specific.  It's just a type with modifiers.  Frankly, a reasonable way to represent it IMO would be like so:

```g4
function_pointer_type
  : 'delegate' '*' ('cdecl' | 'managed' | 'stdcall' | 'thiscall' | 'unmanaged')? '<' type (',' type)* '>'

type
   : /* ... */
   : modified_type;

modified_type
    : modifier* type;
```

333fred replied:
Well, the point is that `ParameterSyntax` technically allows this, from a "this will only error in semantic binding" perspective, but in this grammar at least, that would be an error.

---
In reply to: [348212328](https://github.com/dotnet/roslyn/pull/39879#discussion_r348212328) [](ancestors = 348212328)

CyrusNajmabadi replied:
sorry, i feel like i'm not understanding the point you're making.  want to try gitter/slack?

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -131,7 +131,8 @@ private bool LooksLikeTypeOfPattern(SyntaxKind tk)
                   (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken)) ||
-                LooksLikeTupleArrayType();
+                LooksLikeTupleArrayType() ||
+                IsFunctionPointerStart(); // We'll parse the function pointer, but issue an error in semantic analysis
```
at thsi point, i would just break this huge expression into multiple statements.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -1857,7 +1859,7 @@ private bool IsTypeDeclarationStart()
             {
                 case SyntaxKind.ClassKeyword:
-                case SyntaxKind.DelegateKeyword:
+                case SyntaxKind.DelegateKeyword when PeekToken(1).Kind != SyntaxKind.AsteriskToken:
```
did this just break the switch to be a linear search instead of normal integer-switch-dispatch?

CyrusNajmabadi replied:
technically ever so slightly more work.  but i think it's much clearer to write: `when !IsFunctionPointerStart()`  up to you.

gafter replied:
No, adding `when` clauses doesn't slow things down.

---
In reply to: [348013181](https://github.com/dotnet/roslyn/pull/39879#discussion_r348013181) [](ancestors = 348013181)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -2533,6 +2535,11 @@ private bool IsEndOfNameInExplicitInterface()
+        private bool IsEndOfFunctionPointerParameterList()
+        {
+            return this.CurrentToken.Kind == SyntaxKind.GreaterThanToken;
+        }
```
nit.  `=>`

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -3547,6 +3554,7 @@ private bool IsPossibleParameter()
                 case SyntaxKind.ArgListKeyword:
                 case SyntaxKind.OpenParenToken:   // tuple
+                case SyntaxKind.DelegateKeyword when IsFunctionPointerStart(): // Function pointer type
                     return true;
```
same concern about linear dispatch instead of a jump-table.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5886,63 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                    if (CurrentToken.Kind != SyntaxKind.LessThanToken)
+                    {
+                        _ = EatToken();
+                    }
```
what's going on here.  why do you just consume wahtever token is next?  For example, say the user had `delegate*;`  or `delegate*>` (where the `>` is the terminator of an ourter list). You'd just eat that token unilaterally?

I think this is trying to consume other identifiers in case the user mispells  the calling convention token. In that case, i would specifically check that this is an identifier (or maybe identifier/keyword) before consuming blindly.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5886,63 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+
+            do
+            {
+                switch (CurrentToken.Kind)
```
should you only do this *if* you saw hte `<`.  Otherwise, you're just blindly scanning through a list where there isn't a clear indication you even are in a list.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5886,63 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                }
+
+                _ = EatToken(SyntaxKind.CommaToken);
+            } while (true);
```
while on next line.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5886,63 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                        break;
+                }
+
+                ScanType(out _);
```
1. doc that you don't need these values.
2. does scantype not return something?  if it does, do you want to do `_ =` for consistency?

333fred replied:
For 1, all the `ScanType` functions are pretty consistent here in that they only need the last token. If you open the file in the IDE, for example, you'll see fading from the dead assignment analyzer. We should probably replace a lot of things with `_ =` for clarity.

---
In reply to: [348016168](https://github.com/dotnet/roslyn/pull/39879#discussion_r348016168) [](ancestors = 348016168)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6045,6 +6114,9 @@ bool canBeNullableType()
         }
 
+        private bool IsFunctionPointerStart() => CurrentToken.Kind == SyntaxKind.DelegateKeyword
+                                                 && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
```
nit: `=>` on next line indented.  Then no wrapping of the actual expr.

gafter replied:
Consider `&& peekToken(2).Kind = SyntaxKind.IdentifierToken && peekToken(2).King == SyntaxKind.LessThanToken`.

---
In reply to: [348016475](https://github.com/dotnet/roslyn/pull/39879#discussion_r348016475) [](ancestors = 348016475)

333fred replied:
`delegate*` is already unambiguous, plus I would need to add handling for the calling convention which would not be trivial. I'm going to leave this as is.

---
In reply to: [348134075](https://github.com/dotnet/roslyn/pull/39879#discussion_r348134075) [](ancestors = 348134075,348016475)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    SyntaxKind.StdcallKeyword => EatContextualToken(SyntaxKind.StdcallKeyword),
+                    _ => parseInvalidCallingConvention(),
+                }
+                : null;
```
`var callingConvention = kind != lessthan ? EatCallingConvention() : null;`

Then define `EatCallingConvention` below (as local or other instance method).   Interspersing a large pattern like this just feels difficult to read and grasp.

Also, all your branches appear the same.  Basically, `ConvertToKeyword(this.EatToken())`.  So it feels like a normal switch where those 5 cases call the same thing is clearer and less error prone.

CyrusNajmabadi replied:
you could also just have all the code be in its own local function (since you already do that for parseInvalidCallingConv).  The mixing/matching of styles/logic here isn't great IMO.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                            break;
+                        }
+
+                        PostSkipAction skipBadFunctionPointerParameterListTokens()
```
ick.  can you move this below.  it gets really confusing to read interspersed normal code-flow with helper functions.

333fred replied:
No, it's capturing local state.

---
In reply to: [348019621](https://github.com/dotnet/roslyn/pull/39879#discussion_r348019621) [](ancestors = 348019621)

CyrusNajmabadi replied:
thanks.  now i hate it more :)

333fred replied:
I could put a `continue` above it to make the separation clearer?

---
In reply to: [348237933](https://github.com/dotnet/roslyn/pull/39879#discussion_r348237933) [](ancestors = 348237933)

CyrusNajmabadi replied:
could you instead move it, and just pass the values directly, instead of capturing them?

333fred replied:
Apparently I lied, the variable I was capturing is at the top level and is fine to be on the outer level.

---
In reply to: [348244719](https://github.com/dotnet/roslyn/pull/39879#discussion_r348244719) [](ancestors = 348244719)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                            case SyntaxKind.OutKeyword:
+                            case SyntaxKind.InKeyword:
+                                modifiers.Add(EatToken(CurrentToken.Kind));
+                                break;
```
maybe comment that modifiers will be checked later.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    SyntaxFacts.IsPredefinedType(CurrentToken.Kind))
+                {
+                    return null;
+                }
```
note: i think we should scan and be resilient to people trying to write things like so:

```c#
delegate* ret_type<arg_types> // and
delegate* ret_type(arg_types) // and
delegate* ret_type(parameters) // and
```

We should detect this and give good errors on what it actually expected.  This is similar to how we detect common reorderings of operator pieces since people commonly confuse them.  

Note: i think it's arguable that your syntax model should *actually* be:

```delegate * calling_conv? < parameters >```

You can then validate *later* that the parameters don't have names/initializers.  But it will mean we can reuse existing parsing logic, and it means that the syntax model can accurately represent common mistakes that people *will make*.

333fred replied:
I've added test cases for all of these scenarios, though the `ret_type` part is actually interpreted as an invalid calling convention rather than a return type.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -10994,6 +11176,12 @@ private bool ShouldParseLambdaParameterType(bool hasModifier)
                 return true;
             }
 
+            // If this is a function pointer type
```
comment not useful :)

333fred replied:
I'm just following the rest of the style of the function.

---
In reply to: [348022338](https://github.com/dotnet/roslyn/pull/39879#discussion_r348022338) [](ancestors = 348022338)

333fred replied:
If I need to push another revision, I'll remove this PR. Otherwise, I'll followup later

---
In reply to: [348022338](https://github.com/dotnet/roslyn/pull/39879#discussion_r348022338) [](ancestors = 348022338)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:
```diff
@@ -483,7 +483,7 @@ protected SyntaxToken EatToken(SyntaxKind kind)
 
         // Consume a token if it is the right kind. Otherwise skip a token and replace it with one of the correct kind.
-        protected SyntaxToken EatTokenAsKind(SyntaxKind expected)
+        protected SyntaxToken EatTokenAsKind(SyntaxKind expected, bool reportError = true)
```
i don't see any place (currently) where EatTokenAsKind is called with `reportError: false`.  Is this needed?

CyrusNajmabadi commented:
LGTM.  However, there are several things i think can/should be done to make the parser more production-ready for the types of cases we should expect users to be running into.

gafter commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -223,6 +223,61 @@
+      <Kind Name="ManagedKeyword" />
+      <Kind Name="StdcallKeyword" />
+      <Kind Name="ThiscallKeyword" />
+      <Kind Name="UnmanagedKeyword" />
```
>Kind [](start = 7, length = 4)

I don't think this list belongs in the parser.  I think it belongs in semantic analysis.  The parser should accept any identifier and let semantic analysis sort it out.  I could be wrong about this...

CyrusNajmabadi replied:
IDE would def like nice kinds for contextual keywords fwiw.  If these really are well-known names that the compiler cares about, then this does make sense to me.  I don't feel super strongly about it though.

gafter replied:
Some people don't like "overfitting of the semantic model" ;)

---
In reply to: [348125579](https://github.com/dotnet/roslyn/pull/39879#discussion_r348125579) [](ancestors = 348125579)

333fred replied:
> The parser should accept any identifier and let semantic analysis sort it out.

This is what I've implemented, as I consume any other identifier and replace with a missing token.

---
In reply to: [348138189](https://github.com/dotnet/roslyn/pull/39879#discussion_r348138189) [](ancestors = 348138189,348125579)

CyrusNajmabadi replied:
> Some people don't like "overfitting of the semantic model" ;)

Good point.  I would be ok with this being just an identifier and being purely semantic checked by the compiler.

gafter commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5886,63 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                default:
+                    if (CurrentToken.Kind != SyntaxKind.LessThanToken)
+                    {
+                        _ = EatToken();
```
>EatToken [](start = 28, length = 8)

ScanType is supposed to scan the tokens to see if we have a type.  If we don't have a less-than token, then we don't have a type and this should return an appropriate flag saying so.

333fred replied:
Well, `delegate*` can _only_ mean one thing: someone was trying to write a function pointer. That's why I went with this form.

---
In reply to: [348132938](https://github.com/dotnet/roslyn/pull/39879#discussion_r348132938) [](ancestors = 348132938)

CyrusNajmabadi replied:
I agree with @333fred on that point.  It's an unambiguous prefix for a func-ptr type.  So i think once that part is scanned, it's fine to say that we're still in "we're seeing a type" mode.

333fred commented:
Pushed an updated that addressed a lot of feedback. There will still be a test failure as I haven't updated the AllInOneCSharp file, I will do this later.

333fred commented:
@CyrusNajmabadi @gafter, I've addressed feedback. Please take another look.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+            {
+                return SkipBadTokensWithExpectedKind(isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                                                                      abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
+                                                                      expected: SyntaxKind.CommaToken, trailingTrivia: out _);
```
wonky indent.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            // Invalid calling conventions will be reported at type binding time
+            var callingConvention = IsPossibleCallingConvention() ? EatTokenAsKind(SyntaxKind.IdentifierToken, reportError: false) : null;
```
can't you just do EatToken?  or was there a particular need to have to do AsKind?

333fred replied:
Yes. `EatToken` will return the natural kind of the token, which, if it's a predefined type, will be of the wrong kind. We need to convert anything that isn't an `IdentifierToken` to a missing token, which `EatToken` will not do.

---
In reply to: [348849497](https://github.com/dotnet/roslyn/pull/39879#discussion_r348849497) [](ancestors = 348849497)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                lessThanTokenError = WithAdditionalDiagnostics(lessThanTokenError, GetExpectedTokenError(SyntaxKind.LessThanToken, SyntaxKind.None));
+                var typesError = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+                var identifierError = CreateMissingIdentifierName();
+                identifierError = AddError(identifierError, ErrorCode.ERR_TypeExpected);
```
1. consider just inlining some of this.  i.e. `var identifierError = AddError(CreateMissingIdentifierName(), ErrorCode.ERR_TypeExpected);`
2. doesn't CreateMissingIDentifierName already add an error?  do we get multiple errors?

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
```
i still passionately hate this type.  It's really not in line with most of the rest of the grammar. We barely have any usage of "or" to indicate this-or-that, and it would now be one of the longest names we have despite it being a very simple concept.  Really, these are just types with modifiers.  I honestly think this should just be `ModifiedType` and things will be clear and fine.   We do not have to overfit the grammar here.

333fred replied:
Removed.

---
In reply to: [348850416](https://github.com/dotnet/roslyn/pull/39879#discussion_r348850416) [](ancestors = 348850416)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```
i don't get the need for these two checks.  why is `CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind)` not a sufficient check?

333fred replied:
To try and have a little nicer parse trees. My guess here is that, if there's no list start in sight, we'll have better error recovery if we don't blindly parse this given token as a calling convention. For example, the user could be in the process of adding a cast, something like `(delegate*MyFunctionCall()`, where the end goal is `(delegate*<void>)MyFunctionCall()`.

---
In reply to: [348850685](https://github.com/dotnet/roslyn/pull/39879#discussion_r348850685) [](ancestors = 348850685)

CyrusNajmabadi replied:
i see.  
1. that sounds reasonable.
2. imo we should doc this.  code in the parser that doesn't do the simplest/minimal sort of check possible tends to be difficult to understand later on.  I looked at this and said "i cannot reconcile what this doing against the grammar".  My reading was correct, but your justification was also correct.  The difference between the two is where comments are useful.  Thanks! :)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
+               // Treat predefined types as possible calling conventions, as they're a single
+               // token. This will be a semantic error, but it'll result in a nicer parse tree.
+               (CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind));
```
why only predefined types?  why not any keyword?

333fred replied:
Again in search of nicer parse trees. I could understand putting a predefined type there accidentally because you thought that's where the return type went, but other keywords are almost certainly part of some other construct and it would be better to parse them as they are, rather than consuming that token and attempting to parse that thing. For example:
```cs
delegate* // Declaration in progress
while (true) { }
```
It would be better to just ignore the `while` here, even though the token that follows it _is_ the start of what we could consider a function parameter list start.

---
In reply to: [348850737](https://github.com/dotnet/roslyn/pull/39879#discussion_r348850737) [](ancestors = 348850737)

CyrusNajmabadi replied:
another great reason :)   Can you please comment accordingly?  this sort of knowledge is great to have in the parser so people later on understand what's going on.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
```
why not just ParseModifiers?  then you can flexibly handle user mistakes and report back a standard error for it using our modifier-checking code?

333fred replied:
Because I didn't see that method :)

---
In reply to: [348865046](https://github.com/dotnet/roslyn/pull/39879#discussion_r348865046) [](ancestors = 348865046)

333fred replied:
Note: after investigation, none of the existing ParseXModifiers code paths work quite correctly here, and it's not worth the effort to share code with ParseParameterModifiers here because the majority of that function is just checking to see if language versions are set.a

---
In reply to: [349262113](https://github.com/dotnet/roslyn/pull/39879#discussion_r349262113) [](ancestors = 349262113,348865046)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
```
repeated with the scan code.  consider extracting into IsFunctionPointerModifier

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
```
so... i'm a little nervous about what you're doing her.e are you *sure* it's correct?  specifically... you're allowing yourself to skip here.  but if we skip, i don't see any code that is properly adding the skipped tokens to anything reasonable.  This feels like an invariant bug in the making.  BUt i'm not 100% clear on the patterns the parser uses here... so maybe this is correct?  can you triple check since invariant bugs are awful?

333fred replied:
`SkipBadSeparatedListTokensWithExpectedKind` adds skipped tokens as trailing trivia to the last element of `types`.

---
In reply to: [349283955](https://github.com/dotnet/roslyn/pull/39879#discussion_r349283955) [](ancestors = 349283955)

CyrusNajmabadi replied:
ok thanks.  sorry for the alarm, it just looked really wonky.  since it looks like you always have a single type parsed out, you always have a place to put the skipped tokens.  thanks!

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
```
this feels like a bug right?  or are you really supposed to pass null here and ignore any result?  are all skipped tokens properly added to the `types` collection passed below?

333fred replied:
Yeah, this is really how it's supposed to work. See [http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,f32aaddddf4a6d5b](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,f32aaddddf4a6d5b) as the pre-existing example I copied this from. It definitely does feel like the type of pattern that would not have been written if NRT was a thing since day 1.

---
In reply to: [349284251](https://github.com/dotnet/roslyn/pull/39879#discussion_r349284251) [](ancestors = 349284251)

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
```
note to self: feels like p.IsTerminator should be rolled into the helper function and not be required by all these functions.

CyrusNajmabadi commented:
Parser/Grammar changes LGTM.  THanks!

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1736,6 +1736,12 @@ internal enum ErrorCode
 
+        #region diagnostics introduced in preview
+
+        ERR_InvalidFunctionPointerCallingConvention = 8752,
```
>8752 [](start = 54, length = 4)

Consider leaving some gap (for example starting with 8800) to simplify merges from other branches.

333fred replied:
I'm fine with dealing with the conflicts. Having a gap would annoy me :)

---
In reply to: [349347020](https://github.com/dotnet/roslyn/pull/39879#discussion_r349347020) [](ancestors = 349347020)

AlekseyTs replied:
> I'm fine with dealing with the conflicts. Having a gap would annoy me :)

As long as you are willing to adjust tests to reflect changes in numbers

---
In reply to: [349832744](https://github.com/dotnet/roslyn/pull/39879#discussion_r349832744) [](ancestors = 349832744,349347020)

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -5945,6 +5945,15 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+  <data name="ERR_InvalidFunctionPointerCallingConvention">
+    <value>'{0}' is not a valid calling convention for a function pointer. Valid conventions are 'cdecl', 'managed', 'unmanaged', 'thiscall', and 'stdcall'.</value>
+  </data>
+  <data name="ERR_InvalidFunctionPointerCallingConvention_Title">
```
>ERR_InvalidFunctionPointerCallingConvention_Title [](start = 14, length = 49)

Is this needed for an error?

333fred replied:
It's not needed in this PR directly anymore, but the next step will be binding the type, at which point I'll need to just readd this. So I left it in.

---
In reply to: [349347669](https://github.com/dotnet/roslyn/pull/39879#discussion_r349347669) [](ancestors = 349347669)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+        private ScanTypeFlags ScanFunctionPointerType(out SyntaxToken lastTokenOfType)
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            _ = EatContextualToken(SyntaxKind.DelegateKeyword);
```
>EatContextualToken [](start = 16, length = 18)

Why "Contextual"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
+               // Treat predefined types as possible calling conventions, as they're a single
+               // token. This will be a semantic error, but it'll result in a nicer parse tree.
+               (CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind));
```
>IsPredefinedType [](start = 67, length = 16)

Is there a reason to limit this to just predefined types? Would we want to allow any single "word" (```IsSomeWord```)?

AlekseyTs replied:
This is covered in the comment before the function

---
In reply to: [349358002](https://github.com/dotnet/roslyn/pull/39879#discussion_r349358002) [](ancestors = 349358002)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+        // even though it does look like it could be the start of an invalid function pointer definition.
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```
>PeekToken(1) [](start = 59, length = 12)

Is it always safe to peek one token ahead?

333fred replied:
Yes, I'll add a test. If we're at the end of file, we'll get back more end of file tokesn.

---
In reply to: [349359044](https://github.com/dotnet/roslyn/pull/39879#discussion_r349359044) [](ancestors = 349359044)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+        // even though it does look like it could be the start of an invalid function pointer definition.
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```
>IsPossibleFunctionPointerParameterListStart [](start = 15, length = 43)

Why require LessThanToken after the identifier when we are at the end of the text?

333fred replied:
I feel like parsing the identifier at the end of the file like this is such a corner case that, so long as we don't crash and we produce _something_ halfway reasonable, I'm don't want to do anything else special. I'll add a test.

---
In reply to: [349360246](https://github.com/dotnet/roslyn/pull/39879#discussion_r349360246) [](ancestors = 349360246)

AlekseyTs replied:
> I feel like parsing the identifier at the end of the file like this is such a corner case that, so long as we don't crash and we produce something halfway reasonable, I don't want to do anything else special. I'll add a test.

It looks like there are examples in this file to treats ```EndOfFileToken``` specially to keep detecting certain constructs at the end of the text.  For example in ```IsPossibleAccessorModifier```:
```
            if (token.Kind == SyntaxKind.CloseBraceToken || token.Kind == SyntaxKind.EndOfFileToken)
            {
                // If we see "{ get { } public }
                // then we will think that "public" likely starts an accessor.
                return true;
            }
```

---
In reply to: [349362938](https://github.com/dotnet/roslyn/pull/39879#discussion_r349362938) [](ancestors = 349362938,349360246)

333fred replied:
> It looks like there are examples in this file to treats EndOfFileToken specially to keep detecting certain constructs at the end of the text.

I special case the valid calling conventions. Beyond that, I'm fine with whatever is detected.

---
In reply to: [349380710](https://github.com/dotnet/roslyn/pull/39879#discussion_r349380710) [](ancestors = 349380710,349362938,349360246)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
```
>CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword [](start = 57, length = 47)

Is there some magic around this token?

333fred replied:
`readonly` isn't a valid regular parameter modifier, but it's allowed on `ref readonly` returns, so it needs to be supported here.

---
In reply to: [349361489](https://github.com/dotnet/roslyn/pull/39879#discussion_r349361489) [](ancestors = 349361489)

AlekseyTs replied:
> readonly isn't a valid regular parameter modifier, but it's allowed on ref readonly returns, so it needs to be supported here.

It doesn't have to be preceded by ```ref```? Could you give an example?

---
In reply to: [349363828](https://github.com/dotnet/roslyn/pull/39879#discussion_r349363828) [](ancestors = 349363828,349361489)

333fred replied:
It does indeed have to be preceded by `ref`, but I was planning validating the order and presence of modifiers in the type binding step.

---
In reply to: [349370516](https://github.com/dotnet/roslyn/pull/39879#discussion_r349370516) [](ancestors = 349370516,349363828,349361489)

AlekseyTs replied:
> It does indeed have to be preceded by ref, but I was planning validating the order and presence of modifiers in the type binding step.

Why differ from the way we deal with modifiers in regular signatures. And if this is not different, can we share the logic?

---
In reply to: [349371508](https://github.com/dotnet/roslyn/pull/39879#discussion_r349371508) [](ancestors = 349371508,349370516,349363828,349361489)

333fred replied:
> Why differ from the way we deal with modifiers in regular signatures

This isn't particularly different. The only validation that `ParseParameterModifiers` does at this stage, for example, is language version checks. I could add a flag to that function that allows it to check for `readonly` as well, as it doesn't handle them currently, but I decided that there wasn't a huge benefit to doing this. I can modify it if you think it's really warranted though.

---
In reply to: [349378051](https://github.com/dotnet/roslyn/pull/39879#discussion_r349378051) [](ancestors = 349378051,349371508,349370516,349363828,349361489)

333fred replied:
Followed up offline. This isn't quite the same as a regular parameter modifier because it also allows `ref readonly`, where regular parameters do not.

---
In reply to: [349741647](https://github.com/dotnet/roslyn/pull/39879#discussion_r349741647) [](ancestors = 349741647,349378051,349371508,349370516,349363828,349361489)

333fred replied:
That being said, if we ever want to add new modifiers to regular parameters we'll need to do language version checks, so if we have `ParseParameterModifiers` take a flag and understand this, it would probably good for future proofing.

---
In reply to: [349809401](https://github.com/dotnet/roslyn/pull/39879#discussion_r349809401) [](ancestors = 349809401,349741647,349378051,349371508,349370516,349363828,349361489)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+        // construct, and would produce better errors if parsed separately. The user could have this:
+        //
+        // delegate* // Declaration in progress
+        // while (true) {}
```
>while (true) {} [](start = 11, length = 15)

I think the same argument could be made for predefined types:
```
delegate* // Declaration in progress
int x = 0;
```
Perhaps we should eat any keyword as long as it is on the same line? Which would mean that a predefined type on a different line is not a calling convention.

333fred replied:
That's a fair point, though I do think that, for things like `foreach`, `while`, `if`, etc, it would never be a good idea to eat them here, since they are control flow concepts and could never be thought to have meaning here. I'll look into preventing keyword eating when it's not on the same line, though I'll also need to test that we eat a regular identifier on another line there.

---
In reply to: [349363152](https://github.com/dotnet/roslyn/pull/39879#discussion_r349363152) [](ancestors = 349363152)

CyrusNajmabadi replied:
i like this idea. 👍

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+
+            if (IsPossibleCallingConvention())
+            {
+                _ = EatToken();
```
>_ = EatToken(); [](start = 16, length = 15)

I think the code would be more robust if we adjust lastTokenOfType as we go along, rather than making an assumption that, if we get here, we will not get into the ```if``` below and return early because **the current** implementation of ```IsPossibleCallingConvention``` ensures that ```IsPossibleFunctionPointerParameterListStart(CurrentToken)``` is true after this line.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                return ScanTypeFlags.MustBeType;
+            }
+
+            _ = EatToken(SyntaxKind.LessThanToken);
```
>EatToken(SyntaxKind.LessThanToken) [](start = 16, length = 34)

Is this going to eat ```OpenParenToken``` for which ```IsPossibleFunctionPointerParameterListStart(CurrentToken)``` above returns true?

333fred replied:
No, I should just make this a `EatToken()`

---
In reply to: [349366215](https://github.com/dotnet/roslyn/pull/39879#discussion_r349366215) [](ancestors = 349366215)

333fred replied:
And same with the `EatToken(SyntaxKind.GreatherThanToken)` below

---
In reply to: [349367034](https://github.com/dotnet/roslyn/pull/39879#discussion_r349367034) [](ancestors = 349367034,349366215)

333fred commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+            PostSkipAction skipBadFunctionPointerParameterTokens()
+            {
+                return SkipBadTokensWithExpectedKind(isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                                                     abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
```
>p.CurrentToken.Kind == SyntaxKind.CommaToken || [](start = 73, length = 47)

Note: this is redundant, will remove.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -2533,6 +2535,8 @@ private bool IsEndOfNameInExplicitInterface()
             return this.CurrentToken.Kind == SyntaxKind.DotToken || this.CurrentToken.Kind == SyntaxKind.ColonColonToken;
         }
 
+        private bool IsEndOfFunctionPointerParameterList() => this.CurrentToken.Kind == SyntaxKind.GreaterThanToken;
```
>this.CurrentToken.Kind == SyntaxKind.GreaterThanToken [](start = 62, length = 53)

Should we also recognize ```OpenParenToken```?

333fred replied:
Yes, but only the case where we started with an `OpenParenToken`.

---
In reply to: [349368015](https://github.com/dotnet/roslyn/pull/39879#discussion_r349368015) [](ancestors = 349368015)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+                _termState = saveTerm;
+            }
+
+            lastTokenOfType = EatToken(SyntaxKind.GreaterThanToken);
```
>EatToken(SyntaxKind.GreaterThanToken) [](start = 30, length = 37)

Are we really interested in "fake"/missing tokens returned as ```lastTokenOfType```?

333fred replied:
It looks like the only thing that anyone is interested in the `lastTokenOfType` for is to know whether it's a predefined type, which will of course be false for this scenario.

---
In reply to: [349369399](https://github.com/dotnet/roslyn/pull/39879#discussion_r349369399) [](ancestors = 349369399)

333fred replied:
And a couple of uses to ensure we don't parse `Type??`

---
In reply to: [349370219](https://github.com/dotnet/roslyn/pull/39879#discussion_r349370219) [](ancestors = 349370219,349369399)

AlekseyTs replied:
> It looks like the only thing that anyone is interested in the lastTokenOfType for is to know whether it's a predefined type, which will of course be false for this scenario.

It would be more robust to not rely on the **current** usage elsewhere and sticking only to real tokens in the stream. 

---
In reply to: [349370569](https://github.com/dotnet/roslyn/pull/39879#discussion_r349370569) [](ancestors = 349370569,349370219,349369399)

333fred commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                return funcPtr;
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
```
>EatToken [](start = 32, length = 8)

This also needs to be `EatTokenAsKind`

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:
```diff
@@ -494,7 +494,7 @@ protected SyntaxToken EatTokenAsKind(SyntaxKind expected)
 
-            var replacement = CreateMissingToken(expected, this.CurrentToken.Kind, reportError: true);
+            var replacement = CreateMissingToken(expected, this.CurrentToken.Kind, reportError);
             return AddTrailingSkippedSyntax(replacement, this.EatToken());
```
>AddTrailingSkippedSyntax [](start = 19, length = 24)

It feels strange to have no error in a syntax tree that has skipped tokens. If we want to consume keywords as calling conventions, perhaps we should relax/remove kind restriction on it in the syntax.xml and consume the token normally. Or "reinterpret" the keyword token into an identifier token with the text and position of the keyword without creating any skipped tokens.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            {
+                var lessThanTokenError = WithAdditionalDiagnostics(SyntaxFactory.MissingToken(SyntaxKind.LessThanToken), GetExpectedTokenError(SyntaxKind.LessThanToken, SyntaxKind.None));
+                var typesError = _pool.AllocateSeparated<ModifiedType>();
+                var identifierError = AddError(CreateMissingIdentifierName(), ErrorCode.ERR_TypeExpected);
```
>AddError(CreateMissingIdentifierName(), ErrorCode.ERR_TypeExpected); [](start = 38, length = 68)

We already reported an error above. Is it really helpful to report more errors?

333fred commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
```
>EatToken [](start = 39, length = 8)

This will need to be adjusted to potentially consume a bad token if we started with a bad token.

AlekseyTs replied:
>var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken); [](start = 16, length = 61)

Do we want to consume ```CloseParenToken``` here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                var identifierError = AddError(CreateMissingIdentifierName(), ErrorCode.ERR_TypeExpected);
+                var typeError = SyntaxFactory.ModifiedType(default, identifierError);
+                typesError.Add(typeError);
+                var greaterThanTokenError = WithAdditionalDiagnostics(SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken), GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
```
>GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None) [](start = 127, length = 67)

Also looks like an unnecessary error. What errors are we going to report for ```delegate* >```? Are we going to say ```>``` is expected at the ```>``` character?

333fred replied:
I have a test for this, and yes, we do report a `>` missing character on that token.

---
In reply to: [349376764](https://github.com/dotnet/roslyn/pull/39879#discussion_r349376764) [](ancestors = 349376764)

333fred replied:
I'll remove the extra errors.

---
In reply to: [349377335](https://github.com/dotnet/roslyn/pull/39879#discussion_r349377335) [](ancestors = 349377335,349376764)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -128,10 +128,29 @@ private CSharpSyntaxNode ParseTypeOrPatternForIsOperatorCore()
         /// </summary>
         private bool LooksLikeTypeOfPattern(SyntaxKind tk)
```
>LooksLikeTypeOfPattern [](start = 21, length = 22)

It feels strange that part of this function uses parameter ```tk``` to figure out the answer and the other part uses current position in the token stream. This can confuse consumers thinking that it is enough to pass the ```SyntaxKind``` and the current position doesn't matter. Consider removing the parameter.

AlekseyTs commented:
Done with review pass (iteration 4)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
+
+            if (!IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                return ScanTypeFlags.MustBeType;
```
>ScanTypeFlags.MustBeType [](start = 23, length = 24)

It feels like the situation doesn't quite fit the meaning of the value:
```
/// <summary>
/// Definitely a type name: either a predefined type (int, string, etc.) or an array
/// type (ending with a [] brackets), or a pointer type (ending with *s).
/// </summary>
MustBeType,
```

Also, should we somehow indicate that the type is syntactically incomplete?

333fred replied:
I talked with Neal about this during the initial passes. We use ScanTypeFlags for later syntax disambiguation, and `delegate*` is 100% definitely a type. I'll go through existing usages of the flag to make sure that this doesn't add new ambiguities, and update the documentation for the invalid cases.

---
In reply to: [349718535](https://github.com/dotnet/roslyn/pull/39879#discussion_r349718535) [](ancestors = 349718535)

333fred commented:
@AlekseyTs addressed feedback, thanks!

---
In reply to: [557336082](https://github.com/dotnet/roslyn/pull/39879#issuecomment-557336082) [](ancestors = 557336082)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6357,175 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                var missingType = SyntaxFactory.ModifiedType(default, missingTypeName);
+                missingTypes.Add(missingType);
+                // Handle the simple case of delegate*>. We don't try to deal with any variation of delegate*invalid>, as
+                // we we don't know for sure that the expression isn't a relational with something else.
```
>we we [](start = 19, length = 5)

typo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6357,175 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            // we'd have an error anyway, and it's more likely the user intended for it to be a
+            // function pointer convention than not.
+            // PROTOTYPE(func-ptr): refactor this out into a helper method to share with binding
+            switch (CurrentToken.Text)
```
>Text [](start = 33, length = 4)

Should this be ```Text``` or ```ValueText```?

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs:
```diff
@@ -1,6 +1,6 @@
 
-//#define PARSING_TESTS_DUMP
+#define PARSING_TESTS_DUMP
 
```
The change looks unintentional.

AlekseyTs commented:
Done with review pass (iteration 5)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [558364765](https://github.com/dotnet/roslyn/pull/39879#issuecomment-558364765) [](ancestors = 558364765)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6266,13 +6357,175 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            // we'd have an error anyway, and it's more likely the user intended for it to be a
+            // function pointer convention than not.
+            // PROTOTYPE(func-ptr): refactor this out into a helper method to share with binding
+            switch (CurrentToken.ValueText)
```
>CurrentToken.ValueText [](start = 20, length = 22)

I think this change should be backed by a test that observes the difference in behavior.

333fred replied:
> I think this change should be backed by a test that observes the difference in behavior.

After adding a test, I've gone back to `Text` instead of `ValueText`, as we don't want to treat `@cdecl` as a calling convention, given that it's been escaped.

---
In reply to: [350896749](https://github.com/dotnet/roslyn/pull/39879#discussion_r350896749) [](ancestors = 350896749)

AlekseyTs replied:
> After adding a test, I've gone back to Text instead of ValueText, as we don't want to treat @cdecl as a calling convention, given that it's been escaped.

It feels like we are somewhat inconsistent:
- ```delegate* @cdecl``` - return false
- ```delegate* @cdecl <``` - return true

Perhaps we should ignore all escaped identifiers? I think it would be fine to add a prototype comment and follow up on this later.

---
In reply to: [350976752](https://github.com/dotnet/roslyn/pull/39879#discussion_r350976752) [](ancestors = 350976752,350896749)

333fred replied:
> It feels like we are somewhat inconsistent

I don't think that that's any more inconsistent than a non-escaped regular identifier that isn't a known convention, like `MyType`. We special case things that we know are valid calling conventions, but the convention was specifically escaped here so we treat it as any other invalid identifier.

---
In reply to: [351425780](https://github.com/dotnet/roslyn/pull/39879#discussion_r351425780) [](ancestors = 351425780,350976752,350896749)

AlekseyTs replied:
> I don't think that that's any more inconsistent than a non-escaped regular identifier that isn't a known convention, like MyType. We special case things that we know are valid calling conventions, but the convention was specifically escaped here so we treat it as any other invalid identifier.

Then why do we treat ```@cdecl``` as possibly valid calling convention when it is followed by ```<```. If gramar doesn't allow escaped identifiers, then we should not accept them. 

---
In reply to: [351441102](https://github.com/dotnet/roslyn/pull/39879#discussion_r351441102) [](ancestors = 351441102,351425780,350976752,350896749)

AlekseyTs commented:
>             UsingStatement("delegate* @cdecl>", options: TestOptions.RegularPreview,

Consider adding a test for ```delegate* @cdecl < ...```

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/FunctionPointerTests.cs:815 in 3231dcc. [](commit_id = 3231dcc2e7258cfcc3ca3a26f98c1a95ec9ed635, deletion_comment = False)

333fred commented:
>             UsingStatement("delegate* @cdecl>", options: TestOptions.RegularPreview,

That's in my plan for binding tests, but it's not particularly interesting on a syntax level because it's just like any other invalid calling convention.

---
In reply to: [559191659](https://github.com/dotnet/roslyn/pull/39879#issuecomment-559191659) [](ancestors = 559191659)

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/FunctionPointerTests.cs:815 in 3231dcc. [](commit_id = 3231dcc2e7258cfcc3ca3a26f98c1a95ec9ed635, deletion_comment = False)

AlekseyTs commented:
>             UsingStatement("delegate* @cdecl>", options: TestOptions.RegularPreview,

> but it's not particularly interesting on a syntax level because it's just like any other invalid calling convention.

I find the suggested scenario interesting from the parsing point of view because of the contrast to the behavior with scenario in this test. Please consider adding one.

---
In reply to: [559204507](https://github.com/dotnet/roslyn/pull/39879#issuecomment-559204507) [](ancestors = 559204507,559191659)

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/FunctionPointerTests.cs:815 in 27e8523. [](commit_id = 27e85239b591d73818e9f47129f176aab9b37a40, deletion_comment = False)
